src/open_pulse/pulse_utils.cpp

```diff
@@ -0,0 +1,122 @@
+#include "pulse_utils.hpp"
+#include "zspmv.hpp"
+
+complex_t internal_expect_psi_csr(const py::array_t<complex_t>& data,
+                                  const py::array_t<int>& ind,
+                                  const py::array_t<int>& ptr,
+                                  const py::array_t<complex_t>& vec) {
+    auto data_raw = data.unchecked<1>();
+    auto vec_raw = vec.unchecked<1>();
+    auto ind_raw = ind.unchecked<1>();
+    auto ptr_raw = ptr.unchecked<1>();
+
+    auto nrows = vec.shape(0);
+    complex_t temp, expt = 0;
+
+    for (decltype(nrows) row = 0; row < nrows; row++) {
```

Comment:
why not using `auto` here?

Reply:
I have opted for this paranoic mode approach to be sure I'm using the right type and to be safe under posible pybind changes. The thing is that pybind functions returning sizes or lens are not very well documented and you have to dig into the code to know the return type. For example in this case (`shape()` function to initialize `nrows`), it was a `ssize_t`, but others coming from lists, dicts, etc might be different.